File: leetcode solutions/Python/min stack/minStack.py
class MinStack(object):

    def __init__(self):
        self.stack = []
        self.min = None


    def push(self, x):
        if not self.stack:
            self.stack.append(0)
            self.min = x
        else:
            self.stack.append(x - self.min)
            if x < self.min:
                self.min = x


    def pop(self):
        x = self.stack.pop()
        if x < 0:
           # val = self.min
            self.min = self.min - x
        else:
            #val = x + self.min
            pass
        #return val

    def top(self):
        x = self.stack[-1]
        if x > 0:
            return x + self.min
        else:
            return self.min


    def getMin(self):
        return self.min

File: leetcode solutions/Java/min stack/minStack.java
public class MinStack {

    private Stack<Integer> stack = new Stack<>();
    private Stack<Integer> minStack = new Stack<>();


    public void pop() {
        if (stack.pop().equals(minStack.peek())) {
            minStack.pop();
        }
    }

    public void push(int x) {
        stack.push(x);
        if (minStack.isEmpty() || x <= minStack.peek()) {
            minStack.push(x);
        }

    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }

}